descriptions.go

```diff
@@ -126,11 +131,255 @@ type RequestorInfo struct {
+func (wizard IssueWizard) Choose(conf *Configuration, creds CredentialInfoList) ([]IssueWizardItem, error) {
+	// convert creds slice to map for easy lookup
+	credsmap := map[CredentialTypeIdentifier]struct{}{}
+	for _, cred := range creds {
+		credsmap[cred.Identifier()] = struct{}{}
+	}
+
+	contents := wizard.Contents.Choose(conf, credsmap)
+	if wizard.ExpandDependencies != nil && !*wizard.ExpandDependencies {
+		return contents, nil
+	}
+
+	// Each item in contents refers to a credential type that has dependencies, which may themselves
+	// have dependencies. So each item has a tree of dependencies. We must return a list
+	// containing all dependencies of each item in an executable order, i.e. item n in the
+	// list depends only on items < n in the list. We do this as follows:
+	// - by considering element n in items to be dependendent on element n-1, we join all
+	//   dependency trees into one
+	// - of that tree, starting at the leaf nodes and iterating downwards toward the root,
+	//   we put all items in the result list.
+
+	// Fist reverse the array, as we must start iterating at the root (having no dependants)
+	reversed := make([]IssueWizardItem, 0, len(contents))
+	byID := map[CredentialTypeIdentifier]IssueWizardItem{}
+	skipped := 0
+	for i := len(contents) - 1; i >= 0; i-- {
+		item := contents[i]
+		if item.Credential == nil {
+			skipped++
+			continue
+		}
+		reversed = append(reversed, contents[i])
+		byID[*contents[i].Credential] = contents[i]
+	}
+
+	// Build a map containing per level of the dependency tree the (deduplicated) nodes at that level
+	// (root = level 0, its dependencies = level 1, their dependencies = level 2, etc).
+	deps := credentialDependencies{}
+	bylevel := map[int]map[CredentialTypeIdentifier]struct{}{}
+	for i, item := range reversed {
+		if err := wizardItemVisit(bylevel, i, *item.Credential, conf, deps, credsmap); err != nil {
+			return nil, err
+		}
+	}
+
+	// Scanning horizontally, i.e. per level, we iterate across the tree, putting all
+	// credential types that we come across in the result slice. This starts at the leaf nodes
+	// that have no dependencies, and after that across the intermediate nodes whose dependencies
+	// have been put in the result slice in previous iterations.
+	var result []IssueWizardItem                         // to return
+	resultmap := map[CredentialTypeIdentifier]struct{}{} // to keep track of credentials already put in the result slice
+	for i := len(bylevel) - 1; i >= 0; i-- {
+		for id := range bylevel[i] {
+			if _, ok := resultmap[id]; ok {
+				continue
+			}
+			resultmap[id] = struct{}{}
+			if item, present := byID[id]; present {
+				result = append(result, item)
+			} else {
+				current := id // create copy of loop variable to take address of
+				result = append(result, IssueWizardItem{
+					Type:       IssueWizardItemTypeCredential,
+					Credential: &current,
+				})
+			}
+		}
+	}
+
+	result = append(result, contents[len(contents)-skipped:]...)
```

Comment:
I am planning to enforce in `irma scheme sign` that items that do not have a `Credential` field may only be present at the end of the wizard. This ensures that we don't have to consider the case you mentioned, and is also more logical for end users: it makes little sense to have for example a disclosure in the middle of a bunch of issuances, it would make for better UX to have that at the end.

Reply:
The PR you link to is simply this PR, so I think that is a mistake? But nevertheless, if `irma scheme verify` and `irma scheme sign` validate that this scenario does not happen, then it should be fine.

Reply:
Ah so that is a thing that still must be added to the other PR? Because I don't see it there either. Then it is sufficient to me if all constraints are actually enforced.

Reply:
I am planning to enforce this in the other PR, yes. All validation of correctness of wizard data that is not already present will be added in that PR.